aws-fargate/miniwdl_aws_fargate.py

```diff
@@ -0,0 +1,208 @@
+        else:
+            if cls.efs_id is None:
+                raise RuntimeError(
+                    'EFS filesystem config is missing. Mount the filesystem and run miniwdl in a subdirectory of the '
+                    'mountpoint or set the filesystem ID with "export MINIWDL__AWS_FARGATE__EFS_ID=fs-12345678 '
+                    'MINIWDL__AWS_FARGATE__EFS_MOUNTPOINT=/mnt/efs" or any other miniwdl config facility. '
+                    'Use "aws efs describe-file-systems" to list filesystems.'
+                )
+            logger.info("Mounting %s at %s", cls.efs_id, cls.efs_mountpoint)
+            fs_url = f"{cls.efs_id}.efs.{ecs.clients.ecs.meta.region_name}.amazonaws.com:/"
+            subprocess.run(["sudo", "mount", "-t", "nfs4", "-o",
+                            "nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2",
+                            fs_url, cls.efs_mountpoint])
+
+    @classmethod
+    def detect_resource_limits(cls, cfg: config.Loader, logger: logging.Logger) -> Dict[str, int]:
+        return cls._limits
+
+    def poll_task(
+        self, logger: logging.Logger, task_desc, verbose: bool = False
+    ) -> Optional[int]:
+        res = ecs.clients.ecs.describe_tasks(cluster=task_desc["clusterArn"], tasks=[task_desc["taskArn"]])
+        task_desc = res["tasks"][0]
+        if task_desc["lastStatus"] not in self._observed_states:
+            logger.info("Task %s %s", task_desc["taskArn"], task_desc["lastStatus"])
+            self._observed_states.add(task_desc["lastStatus"])
+        if task_desc["lastStatus"] == "STOPPED" and task_desc.get("stopCode") == "TaskFailedToStart":
+            raise Interrupted(task_desc.get("stoppedReason"))
+        return task_desc.get("containers", [{}])[0].get("exitCode")
+
+    def _run(self, logger: logging.Logger, terminating: Callable[[], bool], command: str) -> int:
+        self._observed_states = set()
+        with open(os.path.join(self.host_dir, "command"), "x") as outfile:
+            outfile.write(command)
+
+        image_tag = self.runtime_values.get("docker", "ubuntu:18.04")
+        if ":" not in image_tag:
+            image_tag += ":latest"
+        logger.info("docker image tag %s", image_tag)
+
+        if os.stat(self.host_dir).st_dev != os.stat(self.efs_mountpoint).st_dev:
+            raise RuntimeError(f"miniwdl run directory is outside EFS mountpoint {self.efs_mountpoint}")
+
+        for host_path, container_path in self.input_file_map.items():
+            real_host_path = os.path.realpath(host_path)
+            host_work_path = os.path.join(self.host_dir, "work/_miniwdl_inputs/0", os.path.basename(real_host_path))
+            os.makedirs(os.path.dirname(host_work_path), exist_ok=True)
+            if os.stat(real_host_path).st_dev == os.stat(os.path.dirname(host_work_path)).st_dev:
+                logger.debug("Linking input %s as %s", real_host_path, host_work_path)
+                os.link(real_host_path, host_work_path)
```

Comment:
The input file hard-linking is interesting. Are we blocked from mounting each one directly in the way that SwarmContainer does? I think the hard-linking could be a fine alternative in general, although I need to think through the Directory case. It would be nice also to have a way to mount them read-only by default, possibly by ro-mounting a directory full of hard links.

Reply:
I'm worried mounting each input won't scale because of the ECS task definition size limit of 32 KB (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-quotas.html). Mounting the directory read-only sounds like a good idea, I'd like to address it in a later PR.

Reply:
Good point -- I'm not sure if the existing Swarm implementation might also hit some limit on the number of bind mounts. I'm open to adopting the hard-linking approach (I think we should try to converge on one approach if possible) but I'd probably also want to look at something to clean them up afterwards, since leaving around multiple hardlinks to the same file can be very confusing for users

Reply:
Filed #6 to track this.